### PR TITLE
SYN-5470 Implement modulo operator for storm

### PR DESCRIPTION
**Pull request by blackout-vtx**
- Fixes SYN-5470
- Update grammar to include modulo operator (%)
- Update AST to handle module
- Update stormtypes.Number to have dunders for mod operations
- Update tests to include modulo checks and precedence WRT other math operations

### CODE REVIEW

File: synapse/lib/stormtypes.py [hunk 1]
```diff
@@ -4808,6 +4808,22 @@ def __pow__(self, othr):
...
+    def __mod__(self, othr):
+        if isinstance(othr, float):
+            othr = s_common.hugenum(othr)
+            return Number(s_common.hugemod(self.value(), othr)[1])
+        elif isinstance(othr, (int, decimal.Decimal)):
+            return Number(s_common.hugemod(self.value(), othr)[1])
+        elif isinstance(othr, Number):
+            return Number(s_common.hugemod(self.value(), othr.value())[1])
+
+        mesg = f"'**' not supported between instance of {self.__class__.__name__} and {othr.__class__.__name__}"
```
Comment (invisig0th) [inline synapse/lib/stormtypes.py:4820]: copy/pasta error ( should be % )

**Review by invisig0th — CHANGES_REQUESTED** [review]

File: synapse/lib/stormtypes.py (hunk 1 above)
Comment (Cisphyx) [inline synapse/lib/stormtypes.py:4820]: need to update operator in the error message

Comment (blackout-vtx) [reply synapse/lib/stormtypes.py:4820]: Good catch, thanks. Fixed.

Comment (blackout-vtx) [reply synapse/lib/stormtypes.py:4820]: Good catch, thanks. Fixed.

File: CHANGELOG.rst [hunk 2]
```diff
@@ -62,6 +62,8 @@ Features and Enhancements
   for changes in ``urllib3``. Remove the pinned requirement for the
   ``requests`` library.
   (`#3164 <https://github.com/vertexproject/synapse/pull/3164>`_)
+- Add support for modulo operations in storm.
```
[the remark this answers is not in the document]
Comment (blackout-vtx) [reply CHANGELOG.rst:65]: Removed per our conversation about the changelog/release process.

File: synapse/lib/storm.lark [hunk 3]
```diff
@@ -430,6 +430,7 @@ EXPRMINUS.2: /(?<!-)-/
 EXPRNEG.2: /-(?![^$:.#(\-]|(?>(?<EXPRNEGRECUR>\(((?>[^()"'`]+|(?&EXPRNEGRECUR)|`(?:[^`\\]|\\.)*`|"(?:[^"\\]|\\.)*"|'''.*?'''|'[^']*'(?!'))*)\))|'''.*?'''|`(?:[^`\\]|\\.)*`|"(?:[^"\\]|\\.)*"|'[^']*'(?!')|[^,)}])*,)/
 EXPRTIMES.2: "*"
 EXPRDIVIDE.2: "/"
+EXPRMODULO.2: "%"
 EXPRPOW.2: "**"
 NUMBER.1: /-?(0+\d*(?=\.)|(?<!\d)[1-9]\d*|0+(?![1-9\.]))(\.\d*)?(?![^\s})*\/+\-=><!~]|(?>(?<NUMBERRECUR>\(((?>[^()"'`]+|(?&NUMBERRECUR)|`(?:[^`\\]|\\.)*`|"(?:[^"\\]|\\.)*"|'''.*?'''|'[^']*'(?!'))*)\))|'''.*?'''|`(?:[^`\\]|\\.)*`|"(?:[^"\\]|\\.)*"|'[^']*'(?!')|[^,)}])*,)/
 HEXNUMBER.1: /-?0x[0-9a-fA-F]+(?![^\s})*\/+\-=><!~]|(?>(?<HEXNUMBERRECUR>\(((?>[^()"'`]+|(?&HEXNUMBERRECUR)|`(?:[^`\\]|\\.)*`|"(?:[^"\\]|\\.)*"|'''.*?'''|'[^']*'(?!'))*)\))|'''.*?'''|`(?:[^`\\]|\\.)*`|"(?:[^"\\]|\\.)*"|'[^']*'(?!')|[^,)}])*,)/
```
Comment (Cisphyx) [inline synapse/lib/storm.lark:436]: definitely not an obvious one because these regexes are gross but the lookahead `[^\s})*\/+\-=><!~]` in these needs '%' added to it to allow a number immediately followed by a '%' to work, like`(3%2)`

File: synapse/lib/parser.py [hunk 4]
```diff
@@ -644,6 +645,7 @@ def massage_vartokn(astinfo, x):
     'exprnot': s_ast.UnaryExprNode,
     'exprunary': s_ast.UnaryExprNode,
     'exprcmp': s_ast.ExprNode,
+    'exprmodulo': s_ast.ExprNode,
```
Comment (Cisphyx) [inline synapse/lib/parser.py:648]: don't need this, it is part of the 'exprproduct' rule in the grammar

File: synapse/lib/storm.lark (hunk 3 above)
Comment (blackout-vtx) [reply synapse/lib/storm.lark:436]: Fixed. Thanks for catching that, it's super non-obvious. I also updated the grammar tests to include the math operators without whitespace so hopefully the next person doesn't make the same mistake.

File: synapse/lib/parser.py (hunk 4 above)
Comment (blackout-vtx) [reply synapse/lib/parser.py:648]: Fixed. I think I have a better understanding of how this works now.

File: synapse/lib/storm.lark [hunk 5]
```diff
@@ -430,8 +430,9 @@ EXPRMINUS.2: /(?<!-)-/
 EXPRNEG.2: /-(?![^$:.#(\-]|(?>(?<EXPRNEGRECUR>\(((?>[^()"'`]+|(?&EXPRNEGRECUR)|`(?:[^`\\]|\\.)*`|"(?:[^"\\]|\\.)*"|'''.*?'''|'[^']*'(?!'))*)\))|'''.*?'''|`(?:[^`\\]|\\.)*`|"(?:[^"\\]|\\.)*"|'[^']*'(?!')|[^,)}])*,)/
 EXPRTIMES.2: "*"
 EXPRDIVIDE.2: "/"
+EXPRMODULO.2: "%"
 EXPRPOW.2: "**"
-NUMBER.1: /-?(0+\d*(?=\.)|(?<!\d)[1-9]\d*|0+(?![1-9\.]))(\.\d*)?(?![^\s})*\/+\-=><!~]|(?>(?<NUMBERRECUR>\(((?>[^()"'`]+|(?&NUMBERRECUR)|`(?:[^`\\]|\\.)*`|"(?:[^"\\]|\\.)*"|'''.*?'''|'[^']*'(?!'))*)\))|'''.*?'''|`(?:[^`\\]|\\.)*`|"(?:[^"\\]|\\.)*"|'[^']*'(?!')|[^,)}])*,)/
+NUMBER.1: /-?(0+\d*(?=\.)|(?<!\d)[1-9]\d*|0+(?![1-9\.]))(\.\d*)?(?![^\s})*\/%+\-=><!~]|(?>(?<NUMBERRECUR>\(((?>[^()"'`]+|(?&NUMBERRECUR)|`(?:[^`\\]|\\.)*`|"(?:[^"\\]|\\.)*"|'''.*?'''|'[^']*'(?!'))*)\))|'''.*?'''|`(?:[^`\\]|\\.)*`|"(?:[^"\\]|\\.)*"|'[^']*'(?!')|[^,)}])*,)/
 HEXNUMBER.1: /-?0x[0-9a-fA-F]+(?![^\s})*\/+\-=><!~]|(?>(?<HEXNUMBERRECUR>\(((?>[^()"'`]+|(?&HEXNUMBERRECUR)|`(?:[^`\\]|\\.)*`|"(?:[^"\\]|\\.)*"|'''.*?'''|'[^']*'(?!'))*)\))|'''.*?'''|`(?:[^`\\]|\\.)*`|"(?:[^"\\]|\\.)*"|'[^']*'(?!')|[^,)}])*,)/
```
Comment (Cisphyx) [inline synapse/lib/storm.lark:436]: hexnumber needs that lookahead adjustment too

Comment (blackout-vtx) [reply synapse/lib/storm.lark]: Thanks! I should have caught this. Fixed.

**Review by invisig0th — APPROVED** [review]

**Review by Cisphyx — APPROVED** [review]
